### PR TITLE
Dogstatsd service checks

### PR DESCRIPTION
**Pull request by wang-arthur**
Add service checks to the dogstatsd server. See also: https://github.com/DataDog/dogstatsd-python/pull/24

### CODE REVIEW

**Comment by remh** [comment]
Let's postpone this for 5.3 as clients are not ready for this yet.

**Comment by wang-arthur** [comment]
@remh, could we include this in the 5.2.0 release? That way, users could start sending dogstatsd service checks as soon as the dogstatsd clients are ready, instead of having to wait for the next agent release.

**Comment by remh** [comment]
@wang-arthur Could you run some profiling/performance comparisons with and without your changes? 

The submit_packets method is pretty performance critical so you should try. A good indicator would be the max throughput of the receiver with multiple clients connected to it.

**Comment by wang-arthur** [comment]
Sure, will do

**Comment by remh** [comment]
If performances are ok then we should be able to ship it with 5.2.0

**Comment by wang-arthur** [comment]
To test the effect of these change on `submit_packets` (primarily this: https://github.com/DataDog/dd-agent/pull/1170/files#diff-2289a9b0435abe312320ea1ddd9b4360R556), I ran a profile on this test on the master branch and on my dogstatsd service checks feature branch (https://github.com/DataDog/dd-agent/blob/master/tests/performance/benchmark_aggregator.py#L17).

### master
trial 1
```
7209175 function calls (7209173 primitive calls) in 52.106 seconds
...
 1    4.099    4.099   52.103   52.103 /home/vagrant/workspace/dd-agent/tests/performance/benchmark_aggregator.py:17(test_dogstatsd_aggregation_perf)
  1200000   14.282    0.000   47.933    0.000 /home/vagrant/workspace/dd-agent/aggregator.py:500(submit_packets)
  1200000   13.708    0.000   23.474    0.000 /home/vagrant/workspace/dd-agent/aggregator.py:612(submit_metric)
...
```
trial 2
```
 7209336 function calls (7209334 primitive calls) in 52.427 seconds
...
	1    4.186    4.186   52.424   52.424 /home/vagrant/workspace/dd-agent/tests/performance/benchmark_aggregator.py:17(test_dogstatsd_aggregation_perf)
  	1200000   14.286    0.000   48.158    0.000 /home/vagrant/workspace/dd-agent/aggregator.py:500(submit_packets)
  	1200000   13.741    0.000   23.573    0.000 /home/vagrant/workspace/dd-agent/aggregator.py:612(submit_metric)
...
```
avg total time: 52.2665s
avg submit_packets time: 14.284s
avg submit_metric time: 13.724s

### service check branch
trial 1
```
7209336 function calls (7209334 primitive calls) in 52.747 seconds
...
          1    4.151    4.151   52.744   52.744 /home/vagrant/workspace/dd-agent/tests/performance/benchmark_aggregator.py:17(test_dogstatsd_aggregation_perf)
	  1200000   14.272    0.000   48.510    0.000 /home/vagrant/workspace/dd-agent/aggregator.py:500(submit_packets)
	  1200000   13.876    0.000   23.777    0.000 /home/vagrant/workspace/dd-agent/aggregator.py:612(submit_metric)
...
```
trial 2
```
7209658 function calls (7209656 primitive calls) in 54.871 seconds
...
        1    4.328    4.328   54.868   54.868 /home/vagrant/workspace/dd-agent/tests/performance/benchmark_aggregator.py:17(test_dogstatsd_aggregation_perf)
	1200000   15.062    0.000   50.453    0.000 /home/vagrant/workspace/dd-agent/aggregator.py:546(submit_packets)
	1200000   14.233    0.000   24.490    0.000 /home/vagrant/workspace/dd-agent/aggregator.py:692(submit_metric)
...
```
avg total time: 53.809s
avg submit_packets time: 14.667s
avg submit_metric time: 14.0495s


### comparison
total time: +2.95%
submit_packets: +2.68%
submit_metric: +2.37% (not sure why this would change)

**Comment by wang-arthur** [comment]
@remh, what do you think?

File: aggregator.py [hunk 1]
```diff
@@ -497,6 +502,47 @@ def parse_event_packet(self, packet):
...
+            if message_delimiter in metadata:
+                meta, message = metadata.rsplit(message_delimiter, 1)
+                service_check['message'] = self._unescape_sc_content(message)
+            else:
+                meta = metadata
+
+            if not meta:
+                return service_check
+
+            meta = unicode(meta)
```
Comment (remh) [inline aggregator.py]: That wouldn't work with UTF-8 encoded packets. Could you have a look into this PR https://github.com/DataDog/dd-agent/pull/1279 and mimic the behavior ? 
cc: @LeoCavaille

Comment (wang-arthur) [inline aggregator.py]: Will do, thanks!

File: aggregator.py [hunk 2]
```diff
@@ -497,6 +502,47 @@ def parse_event_packet(self, packet):
...
+                if m[0] == u'd':
+                    service_check['timestamp'] = float(m[2:])
+                elif m[0] == u'h':
+                    service_check['hostname'] = m[2:]
+                elif m[0] == u'#':
+                    service_check['tags'] = sorted(m[1:].split(u','))
+
+            return service_check
+
+        except IndexError, ValueError:
```
Comment (remh) [inline aggregator.py]: Watchout this won't do what you expect.

By doing this you are assigning a caught IndexError exception to ValueError.

```python
Python 2.7.6 (default, Mar 22 2014, 22:59:56)
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> try:
...     l = []
...     print l[1]
... except IndexError, ValueError:
...     print ValueError
...
list index out of range
```
You want to do:
```python
except (IndexError, ValueError)
``` 

which would do The Right Thing (tm):
```python
Python 2.7.6 (default, Mar 22 2014, 22:59:56)
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> try:
...     l = []
...     print l[1]
... except (IndexError, ValueError):
...     print ValueError
...
<type 'exceptions.ValueError'>
```

File: dogstatsd.py [hunk 3]
```diff
@@ -228,6 +233,31 @@ def submit_events(self, events):
...
+        if self.api_key:
+            params['api_key'] = self.api_key
+
+        url = '/api/v1/check_run?{0}'.format(urlencode(params))
+
+        status = None
+        conn = self.http_conn_cls(self.api_host)
+        try:
+            start_time = time()
+            conn.request(method, url, json.dumps(service_checks), headers)
```
Comment (remh) [inline dogstatsd.py]: Could you use the requests library instead ? We recently switched to it in dogstatsd 
https://github.com/DataDog/dd-agent/pull/1290

**Comment by remh** [comment]
Looks great beside the comments made!
Will merge once they are fixed.

Thanks!

**Comment by wang-arthur** [comment]
Thanks for the review @remh! Will address your comments.

**Comment by wang-arthur** [comment]
Made revisions based on @remh's feedback. Merged @LeoCavaille's utc-8 branch into mine, so this should be merged after that one.

**Comment by LeoCavaille** [comment]
Thanks @wang-arthur, rebased your work and squashed your commits to get one commit that can be applied on the current tip of master.

**Comment by wang-arthur** [comment]
Thansk @LeoCavaille!

**Pull request merged**